crypto/sm2/sm2_crypt.c

```diff
@@ -366,7 +378,12 @@ int sm2_decrypt(const EC_KEY *key,
         goto done;
     }
 
-    if (!EVP_DigestInit(hash, digest)
+    fetched_digest = EVP_MD_fetch(libctx, EVP_MD_name(digest), propq);
```

Comment:
This doesn't look correct to me. The passed digest may already be a "fetched" digest (e.g. if EVP_PKEY_CTRL_MD has been used to set a custom digest). Should the fix instead be in `pkey_sm2_encrypt`, to not use the non-fetched `EVP_sm3()`? Or is it actually the fact that SM2 has not been converted to be provider aware that is at the root of the problem?

Reply:
SM2 is still a legacy key..
EVP_PKEY_CTX_ctrl_str() will go thru the legacy path and call EVP_PKEY_CTX_md() which
calls EVP_get_digestbyname()
it then calls EVP_PKEY_CTX_ctrl() and hits the legacy part of that which calls 
EVP_PKEY_CTRL_MD..

I will change pkey_sm2_encrypt() and pkey_sm2_decrypt() to do the fetch.